src/table/components/virtualized-table/hooks/use-get-hypercube-data-queue.ts

```diff
@@ -0,0 +1,67 @@
+import { useMemo, useRef } from 'react';
+
+export interface AbortablePromise {
+  isCancelled: boolean;
+}
+
+const pageToKey = ({ qLeft, qTop, qWidth, qHeight }: EngineAPI.INxPage) => `${qLeft}-${qTop}-${qWidth}-${qHeight}`;
+
+const useGetHyperCubeDataQueue = (
+  getDataPages: (qPages: EngineAPI.INxPage[]) => Promise<EngineAPI.INxDataPage[]>,
+  handleDataPages: (qDataPages: EngineAPI.INxDataPage[]) => void
+) => {
+  const queued = useRef(new Set<EngineAPI.INxPage>()); // Keep track of all unique pages that should be retrieved
+  const ongoing = useRef(new Set<EngineAPI.INxPage[]>()); // Keep track of ongoing request. Should be aborted if page info or layout is changed
+  const finished = useRef(new Set<string>()); // Keep track of all unqiue pages that have added to the queue
+
+  const queue = useMemo(
+    () => ({
+      enqueue: (page: EngineAPI.INxPage) => {
```

Comment:
so when you run `enqueue` you add to `queued`. Then you pop that into `ongoing`, then eventually you put them in `finished` when the fetching is done?

Reply:
The general progression of state is `queued` => `ongoing` => `finished`. So 3 different states.

So when you call `enqueue` there are 3 cases that when a page should NOT be retrieved. That is if it's already `queued`, `ongoing` or `finished`.

BUT instead of have to check `queued`, `ongoing` and `finished` if a page is in any of those sets. I added it directly to `finished`, this means `finished` actually contains all 3 states and more importantly, when you check if page should be retrieved, all you have to look at is the `finished` set. There is 1 caveat to this and that is that if the request fails, they are removed from `finished`.

So `finished` is a fast way to check if a page should be retrieved or ignored.

`queued` is needed to keep track of pages that should be retrieved.

`ongoing` is needed to abort a promise, so essentially, when a promise is settled (data pages have been retrieved) it should be handled by the `handleDataPages` callback. But only if it has not been aborted. Example when a "operation" should be aborted is if the user changes page OR a new layout recieved.